server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java

```diff
@@ -396,7 +397,10 @@ private final class NodeIdsIterator implements Iterator<String> {
 
             @Override
             public boolean hasNext() {
-                if (nodeIds.hasNext() == false && source == NodeIdSource.DESIRED && shard.primary() && wasThrottled == false) {
+                if (nodeIds.hasNext() == false
+                    && source == NodeIdSource.DESIRED
+                    && useFallback(shard, routingNodes)
+                    && wasThrottled == false) {
```

Comment:
Nit: lets check `wasThrottled == false` before `useFallback(..)`. 
In case if the shard was throttled we would not need to check `useFallback` at all as it might be a bit more expensive with streaming other shards in case of `SEARCH_ONLY` role.